lib/luaxtable.c

```diff
@@ -0,0 +1,224 @@
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+	};
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static void luaxtable_release(void *private);
+
+static inline luaxtable_t *luaxtable_newhook(lua_State *L, const lunatik_class_t *cls, int hook)
+{
+	lunatik_object_t *object = lunatik_newobject(L, cls, sizeof(luaxtable_t));	\
+	luaxtable_t *xtable = (luaxtable_t *)object->private;
+
+	xtable->runtime = lunatik_toruntime(L);
+	xtable->type = hook;
+	lunatik_getobject(xtable->runtime);
+	lunatik_registerobject(L, 1, object);
+	return xtable;
+}
+
+#define LUAXTABLE_NEWHOOK(hook, HOOK)					\
+static int luaxtable_new##hook(lua_State *L) {				\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	lunatik_checkfield(L, 1, "name", LUA_TSTRING);			\
+	const char *name = lua_tostring(L, -1);				\
+									\
+	if (strlen(name) >= XT_EXTENSION_MAXNAMELEN) {			\
+		luaL_error(L, "name too long");				\
+	}								\
+									\
+	lunatik_checkfield(L, 1, "revision", LUA_TNUMBER);		\
+	uint8_t revision = (uint8_t)lua_tointeger(L, -1);		\
+									\
+	lunatik_checkfield(L, 1, "family", LUA_TNUMBER);		\
+	unsigned short family = (unsigned short)lua_tointeger(L, -1);	\
+									\
+	unsigned proto = (unsigned short)lua_tointeger(L, -1);		\
+									\
+	lunatik_checkfield(L, 1, "checkentry", LUA_TFUNCTION);		\
+	lunatik_checkfield(L, 1, "destroy", LUA_TFUNCTION);		\
+	lunatik_checkfield(L, 1, #hook, LUA_TFUNCTION);			\
+									\
+	struct xt_##hook *hook = lunatik_checkalloc(L, sizeof(struct xt_##hook));	\
+	memcpy((void *)hook->name, name, XT_EXTENSION_MAXNAMELEN);	\
+	hook->revision = revision;					\
+	hook->family = family;						\
+	hook->proto = proto;						\
+	hook->hook = luaxtable_##hook;					\
+	hook->checkentry = luaxtable_##hook##_check;			\
+	hook->destroy = luaxtable_##hook##_destroy;			\
+	hook->me = THIS_MODULE;						\
+									\
+	if (xt_register_##hook(hook) != 0) {				\
+		lunatik_free(hook);					\
+		luaL_error(L, "unable to register" #hook);		\
+	}								\
+									\
+	luaxtable_t *xtable = luaxtable_newhook(L, &luaxtable_class, HOOK); 	\
+	xtable->hook = hook;						\
+	return 1;							\
+}
+
+#define LUAXTABLE_RELEASE_HOOK(table, hook)	\
+do {						\
+	xt_unregister_##hook(table->hook);	\
+	lunatik_free(table->hook);		\
+	table->hook = NULL;			\
+} while (0)
+
+static int luaxtable_newmatch(lua_State *L);
+static int luaxtable_newtarget(lua_State *L);
```

Comment:
```suggestion
LUAXTABLE_NEWHOOK(match, LUAXTABLE_TMATCH);
LUAXTABLE_NEWHOOK(target, LUAXTABLE_TTARGET);
```

Reply:
you can also move `luaxtable_*_check` and `luaxtable_*_destroy` to above https://github.com/luainkernel/lunatik/pull/113/files#diff-292f6a11c6f52f91544b0b47343e04b5d6516316ca5a075844659348899709ffR63

Reply:
to be clear, my point here is that it's better to add the prototypes of the "static/fixed" functions than the "dynamic/generated by templates" ones.. you can refer to luanotifier